### PR TITLE
Provide GC tests with isolation for greater reliability

### PR DESCRIPTION
**Pull request by AArnott**
This enables a bunch of GC pressure tests for running on AppVeyor that previously would fail. :smile:

### CODE REVIEW

File: src/IsolatedTestHost/IsolatedTestHost.csproj [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFrameworks>net45</TargetFrameworks>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```
Comment (IlyaBiryukov) [inline src/IsolatedTestHost/IsolatedTestHost.csproj]: Why do we need documentation file for the test host?

File: src/IsolatedTestHost/Program.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    internal static class Program
+    {
+        private static int Main(string[] args)
+        {
+            if (args.Length < 3)
+            {
+                return (int)ExitCodes.CommandLineArgsIncomplete;
+            }
+
+            string assemblyFile = args[0];
```
Comment (IlyaBiryukov) [inline src/IsolatedTestHost/Program.cs]: nit:
```csharp
var (assemblyFile, testClassName, testMethodName) = (args[0], args[1], args[2]);
```

File: src/IsolatedTestHost/Program.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    using System.Reflection;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Xunit;
+
+    internal static class Program
+    {
+        private static int Main(string[] args)
+        {
+            if (args.Length < 3)
```
Comment (IlyaBiryukov) [inline src/IsolatedTestHost/Program.cs]: Check if it is != 3. Or else if you pass extra argument, it's going to be ignored without any indication.

File: src/Microsoft.VisualStudio.Threading.Tests/TestBase.cs [hunk 4]
```diff
@@ -324,5 +325,37 @@ protected void ExecuteOnDispatcher(Func<Task> action)
...
+        /// and GC.
+        /// </summary>
+        /// <param name="testMethodName">The name of the test method.</param>
+        /// <returns>
+        /// A task whose result is <c>true</c> if test execution is already isolated and should therefore proceed with the body of the test,
+        /// or <c>false</c> after the isolated instance of the test has completed execution.
+        /// </returns>
+        /// <exception cref="Xunit.Sdk.XunitException">Thrown if the isolated test result is a Failure.</exception>
+        /// <exception cref="SkipException">Thrown if on a platform that we do not yet support test isolation on.</exception>
+        protected Task<bool> ExecuteInIsolationAsync([CallerMemberName] string testMethodName = null)
```
Comment (IlyaBiryukov) [inline src/Microsoft.VisualStudio.Threading.Tests/TestBase.cs]: Would it make sense to have an xUnit extension, like [Isolated] that would make xUnit run the attributed test method on the isolated host?

File: src/IsolatedTestHost/IsolatedTestHost.csproj (hunk 1 above)
Comment (AArnott) [inline src/IsolatedTestHost/IsolatedTestHost.csproj]: Just so the compiler checks syntax and references in any xml doc comments we define. The warning suppression I also set ensures we aren't obligated to actually write docs.

File: src/Microsoft.VisualStudio.Threading.Tests/TestBase.cs (hunk 4 above)
Comment (AArnott) [inline src/Microsoft.VisualStudio.Threading.Tests/TestBase.cs]: Yes, that would make each usage cleaner, and is what I originally envisioned. But it takes longer to write such support so for now I'm taking the cheaper way out.

File: src/IsolatedTestHost/Program.cs (hunk 2 above)
Comment (AArnott) [inline src/IsolatedTestHost/Program.cs]: I think using deconstruction where it makes sense is good. But... IMO the syntax you propose makes it harder to see the index for each argument.

File: src/IsolatedTestHost/Program.cs (hunk 3 above)
Comment (AArnott) [inline src/IsolatedTestHost/Program.cs]: Done

**Pull request merged**